django/template/defaulttags.py

```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+            espos, eepos = endm.span()
+            if name == partial_name:
+                # Include the full partial definition from opening to closing tag
+                result = full_source[sspos:eepos]
+                break
+            pos = eepos + 1
+        return result
+
+    @property
+    def source(self):
+        template = self.origin.loader.get_template(self.origin.template_name)
+        return self.find_partial_source(template.source, self.name)
+
+    def _render(self, context):
+        return self.nodelist.render(context)
+
+    def render(self, context):
+        "Display stage -- can be called many times"
+        with context.render_context.push_state(self):
+            if context.template is None:
+                with context.bind_template(self):
+                    context.template_name = self.name
+                    return self._render(context)
+            else:
+                return self._render(context)
+
+
+class DefinePartialNode(Node):
+    def __init__(self, partial_name, inline, nodelist):
+        self.partial_name = partial_name
+        self.inline = inline
+        self.nodelist = nodelist
+
+    def render(self, context):
+        """Set content into context and return empty string"""
+        if self.inline:
+            return self.nodelist.render(context)
+        else:
+            return ""
+
+
+class RenderPartialNode(Node):
+    def __init__(self, partial_name, partial_mapping):
+        # Defer lookup of nodelist to runtime.
+        self.partial_name = partial_name
+        self.partial_mapping = partial_mapping
+
+    def render(self, context):
+        return self.partial_mapping[self.partial_name].render(context)
+
+
+@register.tag(name="partialdef")
+def partialdef_func(parser, token):
+    """
+    Declare a partial that can be used later in the template.
+
+    Usage::
+
+        {% partialdef partial_name %}
+        Partial content goes here
+        {% endpartialdef %}
+
+    Stores the nodelist in the context under the key "partial_contents" and can
+    be retrieved using the {% partial %} tag.
+
+    The optional ``inline`` argument will render the contents of the partial
+    where it is defined.
+    """
+    # Parse the tag
+    tokens = token.split_contents()
+
+    # check we have the expected number of tokens before trying to assign them
+    # via indexes
+    if len(tokens) not in (2, 3):
+        raise TemplateSyntaxError(
+            "%r tag requires 2-3 arguments" % token.contents.split()[0]
+        )
+
+    partial_name = tokens[1]
+
+    try:
+        inline = tokens[2]
+    except IndexError:
+        # the inline argument is optional, so fallback to not using it
+        inline = False
+
+    if inline and inline != "inline":
+        warnings.warn(
+            "The 'inline' argument does not have any parameters; "
+            "either use 'inline' or remove it completely.",
+            DeprecationWarning,
+        )
```

Comment:
What does it mean that this raises a deprecation warning? When is the deprecation ending? Why not raise an error here since this is a new feature?

Also we can move this check inside the `if` block right after the `inline = tokens[2]` sentence.

Reply:
Adding to the above, I'm not sure this is tested at all :thinking: